volume/set.go

```diff
@@ -37,24 +38,67 @@ const (
 // Set is a set of volumes.
 type Set struct {
 	volumes   map[string]*Volume
+	providers map[string][]Mount
 	tempDir   string
 	runtime   string
 	volumeDir string
 }
 
+// Provider provides volumes from different sources.
+type Provider interface {
+	// Name of the provider.
+	Name() string
+
+	// CreateVolumesUnder creates volumes under the given directory.
+	CreateVolumesUnder(ctx context.Context, tempDir string) ([]*Volume, error)
+
+	// Delete all resources made by the provider.
+	Delete(ctx context.Context) error
+}
+
 // NewSet returns a new volume set.
 func NewSet(runtime string) *Set {
 	return NewSetWithTempDir(runtime, os.TempDir())
 }
 
 // NewSetWithTempDir returns a new volume set and creates all temporary files under the tempDir.
 func NewSetWithTempDir(runtime, tempDir string) *Set {
-	return &Set{runtime: runtime, volumes: make(map[string]*Volume), tempDir: tempDir}
+	return &Set{
+		runtime:   runtime,
+		providers: make(map[string][]Mount),
+		volumes:   make(map[string]*Volume),
+		tempDir:   tempDir,
+	}
 }
 
 // Add a volume to the set.
 func (vs *Set) Add(v *Volume) {
-	vs.volumes[v.name] = v
+	vs.volumes[fmt.Sprintf("named_%s", v.name)] = v
+}
+
+// AddFrom adds volumes from the given provider.
+func (vs *Set) AddFrom(ctx context.Context, vp Provider) error {
+	if _, exists := vs.providers[vp.Name()]; exists {
+		return fmt.Errorf("failed to add %q: %w", vp.Name(), errdefs.ErrAlreadyExists)
+	}
+
+	volumes, err := vp.CreateVolumesUnder(ctx, vs.tempDir)
```

Comment:
IIUC, when adding volumes from an image, we mount the image at `$VSTempDir/mount`, right? Can you add volumes from 2 containers to the same VS?

Reply:
Ah right. Added one more MkdirTemp